### PR TITLE
Toolset update: VS 2022 17.12 Preview 2, Clang 18.1.8

### LINKED ISSUE #1586: Request LLVM to implement _CountLeadingZeros and _CountLeadingZeros64 when targeting ARM/ARM64
Currently LLVM does not define the MSVC builtins `_CountLeadingZeros` and `_CountLeadingZeros64` when targeting ARM/ARM64. This forces us to maintain a parallel code path in `<bit>` for LLVM targeting ARM.

We should make a bug/feature request asking the LLVM devs to implement these builtins when targeting ARM/ARM64.

### PR DESCRIPTION
**Pull request by StephanTLavavej**
Fixes #1586.

# :scroll: Changelog
- Code cleanups:
  * Removed compiler bug workarounds.
- Infrastructure improvements:
  * Updated dependencies.
    + Updated build compiler to VS 2022 17.12 Preview 2.
    + Updated Clang to 18.1.8 (now required).
    + Updated Python to 3.12.6.

# :memo: Code Format Overhaul
clang-format 18 has noticeably improved its handling of concepts and variable templates. I was able to remove lots of forced wrapping and complete suppression, while needing very few manual adjustments.

As always, the goal is to rely on clang-format to achieve consistency and general readability, not pixel-perfection. I reviewed all of the regen changes, and manually adjusted the ones I thought were obnoxious. Similarly, I removed clang-format suppression when I thought the result was reasonable, even if it was a significant appearance change. We should strive to rely on clang-format for the vast majority of the codebase, giving it only gentle nudges with forced wrapping, and resorting to suppression only when it wants to do something egregious or we need to do something specific for readability or to stay in sync with external sources.

I also explicitly documented my clang-format update process, so we have one less tacit set of steps to perform.

# :gear: Commits

* PowerShell 7.4.5.
* Python 3.12.6.
* New pool.
* VS 2022 17.12 Preview 2. (Still not updating yvals_core.h because the internal toolset has not yet been updated)
* Clang 18.1.8 (we now require Clang 18).
* Remove workarounds for VSO-2170500 (C1XX type trait optimization).
* llvm/llvm-project#59827 was fixed by Clang 18.
* #1586 was fixed by Clang 18.
* Clang 18 understands `-Wno-overriding-option`, so drop `-Wno-unknown-warning-option -Wno-overriding-t-option`.
  + #4452 originally added this.
* Clang 18 understands `-Wno-nan-infinity-disabled`, so drop `-Wno-unknown-warning-option`.
* Add 'TRANSITION, Clang 20' comment noting that `__builtin_isgreater` etc. will be constexpr.
* llvm/llvm-project#62594 was fixed by Clang 18.
* Clang 18 fixed a bug with static call operators.
* Expect clang-format 18.1.8.
* Update .clang-format.
  + Comment my .clang-format update process.
  + Accept defaults, except for:
    - Change BreakAfterAttributes from Leave to Never (the old default).
* Regen clang-format, no manual changes.
* Change AllowBreakBeforeNoexceptSpecifier from Never to OnlyWithParen.
* Regen clang-format, no manual changes.
* LLVM-51935 llvm/llvm-project#51277 was fixed a while ago, but AlignArrayOfStructures appears to be more trouble than it's worth.
* Manual: Add parens to avoid spaces around arrow operator.
* Manual: Detach comment.
* Manual: Add forced wrapping before trailing return types.
* Manual: Remove forced wrapping.
* Manual: Remove lots of clang-format suppression.
* Manual: Remove clang-format suppression and introduce parens.
* Manual: Shrink clang-format suppression in `<concepts>` down to `concept common_with`.
* Comment: Mention `<Windows.h>` with correct capitalization.
* Overhaul header sorting.
  + Use BASIC-style priorities: 10 for yvals/yvals_core, 20 for normal headers, 30 for Windows headers (with fine-grained SortPriority), and 40 for test headers.
  + Mention (yvals|yvals_core) to be more searchable.
  + userenv.h wasn't being included (anymore?).
  + winioctl.h should be spelled lowercase.
  + Make the regex for `__msvc_MEOW.hpp` clearer (we have no other double-underscore headers) and mention it in priority order as there's no chance it'll grab matches first.
  + Add a rule to sort initguid.h first, similar to how winioctl.h sorts last.
* Add newlines to save one column and stay within 120.

# :heart_eyes_cat: Help wanted!

* Clang 18 didn't (completely?) fix the bug affecting `_Zip_transform_constraints` and `_Adjacent_transform_constraints` that we thought was llvm/llvm-project#61763. If anyone can properly reduce and report this, I would be a grateful kitty.

# :white_check_mark: STL-ASan-CI passed
On the first try! https://dev.azure.com/vclibs/STL/_build/results?buildId=17504&view=results

### CODE REVIEW

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move_n/test.cpp [hunk 1]
```diff
@@ -236,7 +236,7 @@ struct memcpy_test {
 
 template <test::ProxyRef IsProxy>
 using test_input  = test::range<test::input, int_wrapper, test::Sized::no, test::CanDifference::no, test::Common::no,
-    test::CanCompare::no, IsProxy>;
+     test::CanCompare::no, IsProxy>;
```
Comment (AlexGuteniev) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move_n/test.cpp:239]: why this extra space?

Comment (StephanTLavavej) [reply tests/std/tests/P0896R4_ranges_alg_uninitialized_move_n/test.cpp:239]: clang-format 18 has interesting new ideas about how to indent wrapped code. I don't entirely understand its new heuristics yet.

File: stl/inc/xmemory [hunk 2]
```diff
@@ -97,7 +97,8 @@ struct _Default_allocate_traits {
 #ifdef __clang__ // Clang and MSVC implement P0784R7 differently; see GH-1532
         _CONSTEXPR20
 #endif // defined(__clang__)
-        void* _Allocate(const size_t _Bytes) {
+        void*
+        _Allocate(const size_t _Bytes) {
```
Comment (CaseyCarter) [inline stl/inc/xmemory:101]: These two occurrences are unfortunate, but not worth blocking the PR over. I'll experiment and see if I can find an improvement.

Comment (StephanTLavavej) [reply stl/inc/xmemory:101]: Agreed, thanks.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by technic** [comment]
Hi, do you maintain a list where it is specified which toolset version requires which minimum clang version? I dont want to get errors like this "Unexpected compiler version, expected Clang 18.0.0 or newer." all of a sudden.

**Comment by CaseyCarter** [comment]
> Hi, do you maintain a list where it is specified which toolset version requires which minimum clang version? I dont want to get errors like this "Unexpected compiler version, expected Clang 18.0.0 or newer." all of a sudden.

We didn't, but I've just scraped the history and added the required Clang version (note that it's always the Clang that ships with Visual Studio) to the notes in https://github.com/microsoft/STL/wiki/Macro-_MSVC_STL_UPDATE, at least going back as far as we have GitHub history.

**Pull request merged**